packet_encoder.go

```diff
@@ -19,9 +19,15 @@ type packetEncoder interface {
 	putInt32Array(in []int32) error
 	putInt64Array(in []int64) error
 
+	// Provide the current offset to record the batch size metric
+	offset() int
+
 	// Stacks, see PushEncoder
 	push(in pushEncoder)
 	pop() error
+
+	// To only record metrics during the real encoder pass
+	doRecordMetrics() bool
```

Comment:
Or even better, have a method `metricsRegistry() *metrics.Registry` that returns the registry or nil depending; this way we can pass the registry in once at the broker and won't have to add it to each request/response struct.